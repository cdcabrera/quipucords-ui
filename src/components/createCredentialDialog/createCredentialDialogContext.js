import { useState } from 'react';
import { AlertVariant } from '@patternfly/react-core';
import { useShallowCompareEffect } from 'react-use';
import { reduxActions, reduxTypes, storeHooks } from '../../redux';
import { useInferredContext as useViewContext } from '../view/viewContext';
import { apiTypes } from '../../constants/apiConstants';
import { translate } from '../i18n/i18n';

/**
 * Return credential dialog response
 *
 * @param {object} options
 * @param {Function} options.useSelector
 * @returns {*}
 */
const useCredential = ({ useSelector: useAliasSelector = storeHooks.reactRedux.useSelector } = {}) => {
  const { credentialType, ...data } = useAliasSelector(({ credentials }) => credentials?.dialog, {});
  const updatedCredentialType = data?.credential?.[apiTypes.API_QUERY_TYPES.CREDENTIAL_TYPE] || credentialType;

  return {
    credentialType: updatedCredentialType,
    ...data
  };
};

/**
 * Return an updated source. Display relative toast messaging after wizard closes.
 *
 * @param {object} options
 * @param {Function} options.addCredential
 * @param {Function} options.t
 * @param {Function} options.updateCredential
 * @param {Function} options.useCredential
 * @param {Function} options.useDispatch
 * @param {Function} options.useViewContext
 * @returns {(function(*, *): void)|*}
 */
const useOnSubmitCredential = ({
  addCredential = reduxActions.credentials.addCredential,
  t = translate,
  updateCredential = reduxActions.credentials.updateCredential,
  useCredential: useAliasCredential = useCredential,
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useViewContext: useAliasViewContext = useViewContext
} = {}) => {
  const { viewId } = useAliasViewContext();
  const [credential, setCredential] = useState();
  const dispatch = useAliasDispatch();
  const { error, fulfilled } = useAliasCredential();

  useShallowCompareEffect(() => {
    if (credential) {
      const { id, ...data } = credential;

      if (id) {
        updateCredential(id, data)(dispatch);
      } else {
        addCredential(data)(dispatch);
      }
    }
  }, [addCredential, credential, dispatch, updateCredential]);

  useShallowCompareEffect(() => {
    if ((credential && fulfilled) || (credential && error)) {
      dispatch([
        {
          type: reduxTypes.credentials.UPDATE_CREDENTIAL_HIDE
        },
        {
          type: reduxTypes.toastNotifications.TOAST_ADD,
          alertType: (error && AlertVariant.danger) || (fulfilled && AlertVariant.success),
          header: t('toast-notifications.title', { context: ['credential', error && 'error'] }),
          message: t('toast-notifications.description', { context: ['credential', error && 'error'] })
        },
        {
          type: reduxTypes.credentials.RESET_ACTIONS
        },
        {
          type: reduxTypes.view.UPDATE_VIEW,
          viewId
        }
      ]);
    }
  }, [credential, dispatch, error, fulfilled, t]);

  return data => setCredential(() => data);
};

/**
 * Functions for handling state updates for credentials.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @returns {{onEdit: Function, onHide: Function, onAdd: Function}}
 */
const useOnUpdateCredential = ({ useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch } = {}) => {
  const dispatch = useAliasDispatch();

  /**
   * On edit show the credential in dialog
   *
   * @event onEdit
   * @param {object|*} credential
   */
  const onEdit = credential => {
    dispatch({
      type: reduxTypes.credentials.EDIT_CREDENTIAL_SHOW,
      credential
    });
  };

  /**
   * Hide the credential dialog
   *
   * @event onHide
   */
  const onHide = () => {
    dispatch({
      type: reduxTypes.credentials.UPDATE_CREDENTIAL_HIDE
    });
  };

  /**
   * on add show the credential dialog
   *
   * @event onAdd
   * @param {string|*} credentialType
   */
  const onAdd = credentialType => {
    dispatch([
      {
        type: reduxTypes.credentials.CREATE_CREDENTIAL_SHOW,
        credentialType
      }
    ]);
  };

  return {
    onEdit,
    onHide,
    onAdd
  };
};

const context = {
  useCredential,
  useOnSubmitCredential,
  useOnUpdateCredential
};

export { context as default, context, useCredential, useOnSubmitCredential, useOnUpdateCredential };
