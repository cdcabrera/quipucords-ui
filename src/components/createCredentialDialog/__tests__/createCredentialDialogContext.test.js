import React from 'react';
import { context, useOnSubmitCredential, useOnUpdateCredential } from '../createCredentialDialogContext';

describe('CredentialsContext', () => {
  it('should return specific properties', () => {
    expect(context).toMatchSnapshot('specific properties');
  });

  it('should attempt to handle submitting a credential with a toast confirmation', async () => {
    const mockCredential = {
      mockId: 'lorem ipsum base id',
      mockName: 'lorem ipsum name'
    };
    const mockUseState = jest.spyOn(React, 'useState').mockImplementation(() => [mockCredential, jest.fn()]);
    const mockDispatch = jest.fn();
    const { result } = await mountHook(() =>
      useOnSubmitCredential({
        useDispatch: () => mockDispatch,
        useCredential: () => ({ fulfilled: true })
      })
    );

    result(mockCredential);

    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch onSubmit');
    mockDispatch.mockClear();
    mockUseState.mockClear();
  });

  it('should handle credential actions for onEdit, onHide, onAdd', async () => {
    const mockDispatch = jest.fn();
    const { result } = await shallowHook(() =>
      useOnUpdateCredential({
        useDispatch: () => mockDispatch
      })
    );

    result.onEdit({
      mockId: 'lorem ipsum base id',
      mockName: 'lorem ipsum name'
    });
    result.onHide();
    result.onAdd('lorem ipsum mock credential type');

    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch onEdit');
    mockDispatch.mockClear();
  });
});
