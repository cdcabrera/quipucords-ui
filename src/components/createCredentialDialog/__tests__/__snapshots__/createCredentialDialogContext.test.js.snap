// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CredentialsContext should attempt to handle submitting a credential with a toast confirmation: dispatch onSubmit 1`] = `
Array [
  Array [
    Object {
      "payload": Promise {},
      "type": "ADD_CREDENTIAL",
    },
  ],
  Array [
    Array [
      Object {
        "type": "UPDATE_CREDENTIAL_HIDE",
      },
      Object {
        "alertType": "success",
        "header": "t(toast-notifications.title, {\\"context\\":\\"credential\\"})",
        "message": "t(toast-notifications.description, {\\"context\\":\\"credential\\"})",
        "type": "TOAST_ADD",
      },
      Object {
        "type": "RESET_CREDENTIAL_ACTIONS",
      },
      Object {
        "type": "UPDATE_VIEW",
        "viewId": undefined,
      },
    ],
  ],
]
`;

exports[`CredentialsContext should handle credential actions for onEdit, onHide, onAdd: dispatch onEdit 1`] = `
Array [
  Array [
    Object {
      "credential": Object {
        "mockId": "lorem ipsum base id",
        "mockName": "lorem ipsum name",
      },
      "type": "EDIT_CREDENTIAL_SHOW",
    },
  ],
  Array [
    Object {
      "type": "UPDATE_CREDENTIAL_HIDE",
    },
  ],
  Array [
    Array [
      Object {
        "credentialType": "lorem ipsum mock credential type",
        "type": "CREATE_CREDENTIAL_SHOW",
      },
    ],
  ],
]
`;

exports[`CredentialsContext should return specific properties: specific properties 1`] = `
Object {
  "useCredential": [Function],
  "useOnSubmitCredential": [Function],
  "useOnUpdateCredential": [Function],
}
`;
